sdk/storage/Azure.Storage.Files.Shares/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,392 @@
+                // later to ensure the file doesn't change while we're
+                // downloading the remaining segments
+                ETag etag = initialResponse.Value.Details.ETag;
+
+                // Download each of the remaining ranges in the file
+                foreach (HttpRange httpRange in GetRanges(initialLength, totalLength))
+                {
+                    // Don't need to worry about 304s here because the ETag
+                    // condition will turn into a 412 and throw a proper
+                    // RequestFailedException
+                    Response<ShareFileDownloadInfo> result = _client.Download(
+                        httpRange,
+                        rangeGetContentHash: false,
+                        conditions,
+                        cancellationToken);
+
+                    // Make sure the ETag for this chunk matches the originally
+                    // recorded one
+                    if (etag != result.GetRawResponse().Headers.ETag)
+                    {
+                        throw new ShareFileModifiedException(
+                            "File has been modified concurrently",
+                            _client.Uri, etag, result.GetRawResponse().Headers.ETag.GetValueOrDefault(), ParseRange(result.Value.Details.ContentRange));
+                    }
+
+                    CopyTo(result.Value, destination, cancellationToken);
+                }
+
+                return initialResponse.GetRawResponse();
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+            finally
+            {
+                scope.Dispose();
+            }
+        }
+
+        private static long ParseRangeTotalLength(string range)
+        {
+            if (range == null)
+            {
+                return 0;
+            }
+            int lengthSeparator = range.IndexOf("/", StringComparison.InvariantCultureIgnoreCase);
+            if (lengthSeparator == -1)
+            {
+                throw ShareErrors.ParsingFullHttpRangeFailed(range);
+            }
+            return long.Parse(range.Substring(lengthSeparator + 1), CultureInfo.InvariantCulture);
+        }
+
+        private static HttpRange ParseRange(string range)
+        {
+            if (range == null)
+            {
+                return new HttpRange(0, 0);
+            }
+            int lengthSeparator = range.IndexOf("/", StringComparison.InvariantCultureIgnoreCase);
+            if (lengthSeparator == -1)
+            {
+                throw ShareErrors.ParsingFullHttpRangeFailed(range);
+            }
+            string[] rangeBounds = range.Substring(6, lengthSeparator - 6).Split('-');
+            return new HttpRange(
+                long.Parse(rangeBounds[0], CultureInfo.InvariantCulture),
+                long.Parse(rangeBounds[1], CultureInfo.InvariantCulture));
+        }
```

Comment:
I'd suspect logic like this be already somewhere in SDK. Could you check and potentially reuse it ? (might need some refactoring in the other parts of sdk).